### PR TITLE
MULE-9503 Add support for non-blocking dispatchers to transports

### PR DESCRIPTION
**Pull request by dfeist**

### CODE REVIEW

File: core/src/main/java/org/mule/processor/EndpointTransactionalInterceptingMessageProcessor.java [hunk 1]
```diff
@@ -36,7 +38,7 @@ public MuleEvent process(final MuleEvent event) throws MuleException
         {
             return event;
         }
-        else
+        else if(TransactionCoordination.getInstance().getTransaction() != null)
```
Comment (pablokraan) [inline core/src/main/java/org/mule/processor/EndpointTransactionalInterceptingMessageProcessor.java]: I'm not sure about this, but seems to me that if EndpointTransactionalInterceptingMessageProcessor is used, then is because there is a transaction in place. If my assumption is correct, the right change would be to avoid using this class.

Comment (dfeist) [inline core/src/main/java/org/mule/processor/EndpointTransactionalInterceptingMessageProcessor.java]: This chain that includes processor is constructed once for the endpoint but if a composite-source is being used, transactions may or may not be in used for a specific request, that is why this is needed here.

File: core/src/main/java/org/mule/execution/MessageProcessorNotificationExecutionInterceptor.java [hunk 2]
```diff
@@ -120,7 +120,7 @@ public void processExceptionReplyTo(MessagingException exception, Object replyTo
         }
         finally
         {
-            if (!NonBlockingVoidMuleEvent.getInstance().equals(result) && fireNotification)
+            if (!(result instanceof VoidMuleEvent) && fireNotification)
```
Comment (pablokraan) [inline core/src/main/java/org/mule/execution/MessageProcessorNotificationExecutionInterceptor.java]: Why is this change required? Was not possible to receive a VoidMuleEvent before this PR?

Comment (dfeist) [inline core/src/main/java/org/mule/execution/MessageProcessorNotificationExecutionInterceptor.java]: This now matches sub-clasess of VoidMuleEvent (such as NonBlockingVoidEvent) which it didn't before.  This wasn't required before as endpoints didn't support non-blocking.  The other updates like this are for the same reason.

File: core/src/main/java/org/mule/transport/AbstractMessageDispatcher.java [hunk 3]
```diff
@@ -213,4 +255,88 @@ protected void applyResponseTransformers(MuleEvent event) throws MuleException
     protected abstract void doDispatch(MuleEvent event) throws Exception;
 
     protected abstract MuleMessage doSend(MuleEvent event) throws Exception;
+
+    protected void doSendNonBlocking(MuleEvent event, CompletionHandler<MuleMessage, Exception> completionHandler)
+    {
+        throw new IllegalStateException("This MessageDispatcher does not support non-blocking");
+    }
+
+    private class NonBlockingSendCompletionHandler implements CompletionHandler<MuleMessage, Exception>
```
Comment (pablokraan) [inline core/src/main/java/org/mule/transport/AbstractMessageDispatcher.java]: Would be better to have this class a a top level element and have it properly tested

File: core/src/test/java/org/mule/endpoint/outbound/OutboundEndpointTestCase.java [hunk 4]
```diff
@@ -80,12 +89,26 @@ public void testDefaultFlowSync() throws Exception
...
+    {
+        OutboundEndpoint endpoint = createOutboundEndpoint(null, null, null, null, MessageExchangePattern.REQUEST_RESPONSE, null);
+
+        SensingNullReplyToHandler nullReplyToHandler = new SensingNullReplyToHandler();
+        MuleEvent event = getNonBlockingTestEventUsingFlow(TEST_MESSAGE, nullReplyToHandler);
+
+        MuleEvent response = endpoint.process(event);
+        assertThat(response, CoreMatchers.<MuleEvent>equalTo(NonBlockingVoidMuleEvent.getInstance()));
+
+        assertThat(getNonBlockingResponse(nullReplyToHandler), CoreMatchers.equalTo(event));
```
Comment (pablokraan) [inline core/src/test/java/org/mule/endpoint/outbound/OutboundEndpointTestCase.java]: Use static imports

File: core/src/test/java/org/mule/endpoint/outbound/OutboundNotificationMessageProcessorTestCase.java [hunk 5]
```diff
@@ -66,4 +66,26 @@ public void testSend() throws Exception
...
     }
 
+    @Test
+    public void testSendNonBlocking() throws Exception
+    {
+        TestEndpointMessageNotificationListener listener = new TestEndpointMessageNotificationListener();
+        muleContext.registerListener(listener);
+
+        OutboundEndpoint endpoint = createTestOutboundEndpoint(null, null, null, null,
+                                                               MessageExchangePattern.REQUEST_RESPONSE, null);
```
Comment (pablokraan) [inline core/src/test/java/org/mule/endpoint/outbound/OutboundNotificationMessageProcessorTestCase.java]: always use static imports...

File: core/src/test/java/org/mule/endpoint/outbound/OutboundNotificationMessageProcessorTestCase.java [hunk 6]
```diff
@@ -66,4 +66,26 @@ public void testSend() throws Exception
...
+        MessageProcessor mp = new OutboundNotificationMessageProcessor(endpoint);
+        SensingNullReplyToHandler nullReplyToHandler = new SensingNullReplyToHandler();
+        MuleEvent event = getNonBlockingTestEventUsingFlow(TEST_MESSAGE, nullReplyToHandler);
+        mp.process(event);
+
+        assertTrue(listener.latch.await(RECEIVE_TIMEOUT, TimeUnit.MILLISECONDS));
```
Comment (pablokraan) [inline core/src/test/java/org/mule/endpoint/outbound/OutboundNotificationMessageProcessorTestCase.java]: Use assertThat

File: core/src/test/java/org/mule/processor/TestNonBlockingProcessor.java [hunk 7]
```diff
@@ -4,7 +4,7 @@
  * license, a copy of which has been included with this distribution in the
  * LICENSE.txt file.
  */
-package org.mule.tck.functional;
+package org.mule.processor;
```
Comment (pablokraan) [inline core/src/test/java/org/mule/processor/TestNonBlockingProcessor.java]: Why is this change of package required?

Comment (dfeist) [inline core/src/test/java/org/mule/processor/TestNonBlockingProcessor.java]: As I'm now using this in mule-core.

File: core/src/test/java/org/mule/tck/junit4/AbstractMuleContextTestCase.java [hunk 8]
```diff
@@ -611,4 +617,25 @@ protected File getFileInsideWorkingDirectory(String fileName)
     {
         return new File(getWorkingDirectory(), fileName);
     }
+
+    protected MuleEvent getNonBlockingTestEventUsingFlow(Object payload, ReplyToHandler replyToHandler) throws Exception
```
Comment (pablokraan) [inline core/src/test/java/org/mule/tck/junit4/AbstractMuleContextTestCase.java]: When merging this to mule4.x, these methods should be on TestEventBuilder

File: core/src/test/java/org/mule/tck/testmodels/mule/TestMessageDispatcher.java [hunk 9]
```diff
@@ -73,4 +95,10 @@ protected void doStop()
     {
         // no op
     }
+
+    @Override
+    protected boolean isSupportsNonBlocking()
```
Comment (pablokraan) [inline core/src/test/java/org/mule/tck/testmodels/mule/TestMessageDispatcher.java]: Is not better to have a NonBlockingTestMessageDispatcher extending from TestMessageDispatcher? (asking as this class is used in some tests that would expect this method to return false)

File: core/src/main/java/org/mule/transport/AbstractMessageDispatcher.java (hunk 3 above)
Comment (dfeist) [inline core/src/main/java/org/mule/transport/AbstractMessageDispatcher.java]: I'd prefer to maintain this a private inner class.  I'll look at adding some more test tests to cover more of this though..

File: core/src/test/java/org/mule/tck/junit4/AbstractMuleContextTestCase.java (hunk 8 above)
Comment (dfeist) [inline core/src/test/java/org/mule/tck/junit4/AbstractMuleContextTestCase.java]: This PR won't be merged to Mule 4, as Mule 4 has no transports.

File: core/src/test/java/org/mule/tck/testmodels/mule/TestMessageDispatcher.java (hunk 9 above)
Comment (dfeist) [inline core/src/test/java/org/mule/tck/testmodels/mule/TestMessageDispatcher.java]: I spent a while moving to the approach you suggest but it just complicates functional testing, and doesn't add much value and there are no existing tests that depend on this.  We currently also use a single dispatcher for testing success/failure caes, rather than two different dispatchers, this is similar.
